### PR TITLE
Restoring drivers with the structural changes.

### PR DESCRIPTION
**Pull request by BDisp**
## Fixes

- Fixes #_____

## Proposed Changes/Todos

- [ ] Todo 1

## Pull Request checklist:

- [ ] I've named my PR in the form of "Fixes #issue. Terse description."
- [ ] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [ ] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [ ] I ran `dotnet test` before commit
- [ ] I have made corresponding changes to the API documentation (using `///` style comments)
- [ ] My changes generate no new warnings
- [ ] I have checked my code and corrected any poor grammar or misspellings
- [ ] I conducted basic QA to assure all features are working

### CODE REVIEW

**Comment by tznind** [comment]
This is going to cause a lot of conflicts with #3791 :(

**Comment by BDisp** [comment]
> This is going to cause a lot of conflicts with #3791 :(

You can't really merge with your PR. This have a lot of fixes and it's very closer with the v2_develop branch? Also have the structural refactoring that @tig did. All the response stuff was removed.

**Comment by BDisp** [comment]
See his requests here https://github.com/gui-cs/Terminal.Gui/pull/3768#issuecomment-2490060423.

**Comment by BDisp** [comment]
`CursesDriver` has an issue with the `Force16Colors` that also happens in the v2_develop. Did you already see this @tig?

WindowsTerminal_whokTx0eEl

**Comment by BDisp** [comment]
`SetupFakeDriver` still having issue. I'll try look at it.

```
  Passed Terminal.Gui.LayoutTests.DimAutoTests.With_Subview_And_Subview_Using_PosAnchorEnd(minWidth: 0, maxWidth: 0, minHeight: 0, maxHeight: 0, expectedWidth: 0, expectedHeight: 0) [< 1 ms]
  Failed Terminal.Gui.LayoutTests.DimAutoTests.With_Subview_Using_PosFunc [1 ms]
  Error Message:
   Assert.Equal() Failure: Values differ
Expected: 25
Actual:   20
  Stack Trace:
     at Terminal.Gui.LayoutTests.DimAutoTests.With_Subview_Using_PosFunc() in D:\a\Terminal.Gui\Terminal.Gui\UnitTests\View\Layout\Dim.AutoTests.PosTypes.cs:line 647
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)

Test Run Failed.
Total tests: 4895
     Passed: 4886
     Failed: 1
    Skipped: 8
 Total time: 1.2931 Minutes
```

**Comment by tig** [comment]
> `SetupFakeDriver` still having issue. I'll try look at it.
> 
> ```
>   Passed Terminal.Gui.LayoutTests.DimAutoTests.With_Subview_And_Subview_Using_PosAnchorEnd(minWidth: 0, maxWidth: 0, minHeight: 0, maxHeight: 0, expectedWidth: 0, expectedHeight: 0) [< 1 ms]
>   Failed Terminal.Gui.LayoutTests.DimAutoTests.With_Subview_Using_PosFunc [1 ms]
>   Error Message:
>    Assert.Equal() Failure: Values differ
> Expected: 25
> Actual:   20
>   Stack Trace:
>      at Terminal.Gui.LayoutTests.DimAutoTests.With_Subview_Using_PosFunc() in D:\a\Terminal.Gui\Terminal.Gui\UnitTests\View\Layout\Dim.AutoTests.PosTypes.cs:line 647
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
>    at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
> 
> Test Run Failed.
> Total tests: 4895
>      Passed: 4886
>      Failed: 1
>     Skipped: 8
>  Total time: 1.2931 Minutes
> ```

I think I fixed this in #3841.

**Comment by BDisp** [comment]
> I think I fixed this in #3841.

But I think the failure also happened after I merged your PR.

**Comment by tznind** [comment]
I have created a PR into this @BDisp that adds IConsoleDriver interface

https://github.com/BDisp/Terminal.Gui/pull/209

Fixes #3846

**Comment by BDisp** [comment]
> I have created a PR into this @BDisp that adds IConsoleDriver interface
> 
> [BDisp#209](https://github.com/BDisp/Terminal.Gui/pull/209)
> 
> Fixes #3846

Thanks for that, I'll take a look. Does it also not going to cause conflicts with #3791 anymore?

**Comment by tznind** [comment]
There will still be conflicts with ansi-parser but adding interface will not substantially increase the number of conflicts.  The two are seperate issues.

**Comment by BDisp** [comment]
An abstract class doesn't act like an interface? The difference is that an abstract can have common code on it and the interface only can have declarations. Can you explain better because what I see is duplicated declarations on both files, `ConsoleDriver` and `IConsoleDriver`. But for sure it's escape me something. Thanks.

**Comment by tznind** [comment]
> An abstract class doesn't act like an interface? The difference is that an abstract can have common code on it and the interface only can have declarations. Can you explain better because what I see is duplicated declarations on both files, `ConsoleDriver` and `IConsoleDriver`. But for sure it's escape me something. Thanks.

Interface allows me to have 'composition' instead of inheritance

This is my new 'driver', I have split everything up into classes.  In order to work with all the classes views etc in Terminal.Gui I have ok like a driver without actually being one.  I will write a 'facade' that implements IConsoleDriver and returns from properties etc all the bits out of my larger network of classes.

389207793-8d9f790c-6c74-49fb-aa67-6530d791fcce

**Comment by BDisp** [comment]
I am more than clear. Thank you.

**Comment by tig** [comment]
> `CursesDriver` has an issue with the `Force16Colors` that also happens in the v2_develop. Did you already see this @tig?
> 
> ![WindowsTerminal_whokTx0eEl](https://private-user-images.githubusercontent.com/13117724/389212113-063ac302-c4c1-4603-9d49-6760f73786a7.gif?

nope.

**Comment by tig** [comment]
> > nope.
> 
> I confirm that this issue doesn't repro anymore and I really don't know how it was fixed because it also was reproducing in the v2_develop. Fortunately it's good now. Thanks for the feedback.

It's possible that it's still there and just requires the unit tests to run in a particular order to cause. We have many unit tests that set static variables on `Application` and `ConfigurationManager` and don't call `Applicaton.ResetState()` or `CM.Reset` which cause these sort of issues.

**Comment by tig** [comment]
Is this PR ready for review/merge?

File: Terminal.Gui/Application/MainLoop.cs
Comment (tig) [inline Terminal.Gui/Application/MainLoop.cs:40]: Why does this need to be `public`?

File: Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: I do not think this should be `public`. Why did you change it?

File: Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs
Comment (tig) [inline Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:62]: Same. Why `public`?

**Review by tig — CHANGES_REQUESTED** [review]
Is it possible to make `IConsoleDriver` `internal` and not `public`.

I feel pretty strongly that "ConsoleDriver" is an IMPLEMENTATION detail of `Application` and `View` should NOT be exposed as a public API.

I worked hard to fix about 80% of the cases where built-in Views and were using `ConsoleDriver` directly. I want to make it 100% before we ship v2.

**Comment by tig** [comment]
Thank you 10000x for doing this. I really appreciate it. See my review comments.

File: Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs
Comment (tznind) [reply Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: This is because i added IConsoleDriver

With a public interface you cannot have internal/private members when implementing.

Question is:
- Are end users allowed to write new drivers?
- Do all drivers need to have the concept of a Clip 

If yes to both then it must be public.

If Clip is used only within the ConsoleDriver class (not even Application) then it can be removed from interface.

File: Terminal.Gui/Application/MainLoop.cs
Comment (tznind) [reply Terminal.Gui/Application/MainLoop.cs:40]: Again this is because it is a return on a member of IConsoleDriver

File: Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs
Comment (tznind) [reply Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: This is a first pass at IConsoleDriver, I am trying to cut it down in scale as part of v2 driver work - but need to start somewhere.

This pr has already diverged a lot from my v2 branch which came off ansi-parser.

Putting interface in both branches is attempt to make merge job easier.

Already I think I will have to throw away my parser integration with these existing drivers because it is too hard to merge with the structural changes

Comment (tig) [reply Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: I do not forsee anyone outside of maintainers building another driver. In the 6+ years I've been working on this project no one has even suggested doing so. So no.

I'm ok with the interface being public, but we need to ensure `Application/View` have all the APIs needed so end-devs don't keep using driver fn directly.

Comment (tznind) [reply Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: Agreed, I envisage this interface shrinking and maybe even being split into sub interfaces later but it is a journey and has to start with small steps.

As soon as a member on interface goes then associated return types eg MainLoop can go back to internal

Comment (tznind) [reply Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: > I do not forsee anyone outside of maintainers building another driver. In the 6+ years I've been working on this project no one has even suggested doing so. So no.

There was  https://github.com/Blazor-Console/HACC

Also can think of lots of potential use cases 

I want to subclass ConsoleDriver but override rendering so it goes out to a file and input from a file - that way I can do automated testing.

I want to Moq console driver so I can test my new views are behaving as expected

I want to subclass ConsoleDriver but override how colors are picked so to avoid combinations that are difficult for colorblind users.

Now maybe these use cases are not tied to ConsoleDriver in the end but to IRender or IInput or something. But only way we get there is incrementally.

A big issue now is how tied Application is to ConsoleDriver and how static everything in Application is - this is next on my explore list.

File: Terminal.Gui/Application/MainLoop.cs
Comment (BDisp) [reply Terminal.Gui/Application/MainLoop.cs:40]: I understand why @tznind needed to do that. When I was working on #3768 to leverage centralize common code I also needed to made `MainLoop` class to public because I needed to add it to the `ConsoleDriver`, which it's a public abstract class. Sometime it's the price to pay for to have centralized automatizations for all drivers.

File: Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs
Comment (tig) [reply Terminal.Gui/ConsoleDrivers/ConsoleDriver.cs:53]: You're on the right track. My concerns are secondary. Keep going!

**Pull request merged**
